### ISSUE TITLE
Review suggestions

### ISSUE THREAD

**Issue by mcm1957** [Trixx34/ioBroker.soliscloud#8]
First of all - THANK YOU for the time and effort you spend to write this adapter.

I would like to give some (preliminary) feedback based on my personal oppinion. This is NOT an offical review and @Apollon77 might have several additional suggestions or even a different oppinion to one or the other statement. 

- [x] #9

  Currently you create all states directly at the instance level. I would suggest to add the plantId as first level. Why? 
  Currently this adapter will support exactly one platId. But its not unrealistic, that a user has more than one system. To avoid installing multiple instances, it looks reasonable to support more than one plant per instance. While adapting the config (using a table) and implementing the necessery code can be done at any time in futuer, changing the state structure would effect users and require changings scripts or even loosing history data. So I thgink its a good idea to create a state structure which can handle this (in future). So I would name the state myinstace.battery_perrcent bette myinstace.1234567890.battery_percent.  

- [x] #10
  
  You configuration seems to contain passwords (or other sensitive information) which is not stored encrypted. As long as those information is not stored inside a table you can encrypt and secure this infomation simply by adding the following configuration to io-package.json. The iob runtime will encrypt / decrypt the data as required; no addition programming effort is required. Note that users will need to reenter this info one time after adding encryption, so drop a note at release notes.
  ```
  "encryptedNative": [
    "password",
    "data2"
  ],
  "protectedNative": [
    "password",
    "data2"
  ],
  ```
- [x] #11
  in case of an error it will be most likely help to know the detailled erro message. So I suggest to log it too. See   "engines": {
    "node": ">= 14"
  },

- [x] #12

  instead of staying insiode an unlimited while loop please consider using this.setInterval or this.setTimeout which call a routine at intervals or after a specific time. Do not forget to use bind().
  see https://github.com/Trixx34/ioBroker.soliscloud/blob/494ed174b658301dd56a4f5952b13cdbd3fcb933/src/main.ts#L131

- [x] state-Id naming (non blocking)
  state ids are typicalle all lowercase (ev. using - and_) or mixed case. '_' and mixed case is unusual. Besides some state are mixed and some are not. But technical is your naming valid, so can keep it, if you prefer too of course.
  
- [x] unused onStateChange handler

  You configred an onStateChange handler but it looks like this handler does not do any important tasks. Please remove the handler if the adapter does not react on state changes.

- [x] reevaluate state roles

  Only the values specified here (https://github.com/ioBroker/ioBroker.docs/blob/master/docs/en/dev/stateroles.md) may be used as state roles. Do not invent new names as this might disturb the functionalyity of other adapters or vis.

  In addition the roles MUST match the read/write functionality. As an example a role value.* requires that the state is a readable state. Input states (write only) should have roles like level.*. Please read the description carefully. States with role 'button' should (must) have attribute 'common.read=false' set. A button ( "Taster" in german only triggers when you press but else has no state), so it should also be not readable. This is also like HomeMatic does it. A switch has clear states true or false and so can be read.

  Please avoid using general roles (state, value) wehnever a dedicated role exists.

  role indicator is definitly incorrect for most states.

- [x]  define units
  
  define units for all states where this is possible (kW, %, ...)

- [x] some labels seem to have no translation into all languages
  
  It seems that some labels are not translated into all supported languages, i.e. ...
  Please consider adding missing translations.

- [x] clear all timeouts / intervals
  all timeouts / intervalls must be cleared at onUnload.
  Failing to do so, will likely break compact mode.

- [x] consider using adapter.setTimeout / this.setTimeout instead of (standard) setTimeout
  
  The adapter package provides wrapper routines for native setTimeout, setInterval, clearTimeout and clearInterval. Using those routines ensures that timers are cancelled on on load. Additional checks on thomse limits might be performed, too. So consider replacing native setTimeout/clearTimeout by adapter.setTimeout/adapter.clearTimeout or this.setTimeout/this.clearTimeout. The same refers to setInterval/clearInterval.
  
- [x] check and limit configurable timeouts / intervals
  
  Node setTimeout/setInterval routines have a maximum allowed value of 2,147,483,647 ms. Using delays larger than 2,147,483,647 ms (about 24.8 days) result in the timeout being executed immediately. So all (user configurable) values passed to setTimeout / setInterval should be checked and limited.
  
- [x] check and limit intervals used to access cloud services
  
  Limit all intervals used to access cloud services to a reasonable value. If users configure an inapprobiate access rate this could lead to suspending the service by the provider for all ioBroker users.  
  

- [x] check and adapt testing
  
  Standard iobroker testing is required but seems to be missing. Please add and setup ./gizthub/workflows/test-and-release.yml
  
- [x] #14
  I strongly suggest to readd and active the deploy functionlity to test-and-release unless you have a special reason to use manual npm deploy. Pleas let me know If I should provide a PR to fix this.
  
- [x] #13
  I strongly suggest to add a timeout to all axios calls. Might be fixed. Axios default is no timeout - so this could lead to a hanging adapter if netzwork / cloud trobles exist.

- [x] add debug log for axios
  I suggest to add debug log stating what request is sent (ev. remove senstive infomation like passowrds) and what response is received from cloud. This should help in case of problems at user systems.

- [x] do not log to console.*, 
  always log to iobrokers log.*. You will not see the output of cosole.log at a running instance and hence a user could not report the erroro occured. see https://github.com/Trixx34/ioBroker.soliscloud/blob/494ed174b658301dd56a4f5952b13cdbd3fcb933/src/lib/apiHelper.ts#L51

- [x] add .releaseconfig
  please add .releaseconfig. @alcalzone/release-script will not correctly function without it

In general I would like to tell you, that this adpater was written in an excellent quality. I've seen worse code from "experienced" developers. Please do not see my remarks as critics. Many of them are pure suggestions, others are simply not documented properly - a fact that must be worked pon by us.

Please feel free at any time if you do not undersatnd a suggestion or need help with it

Thanks for reading and evaluating this suggestions.
McM1957

**Comment by Trixx34** [comment]
Hi!

Thanks for taking the time to review my code and provide this level
of feedback. I'll do my best to make sure the adapter is as compliant as
possible to the guidelines.
I'll see if I can improve on some of these points this evening or in the
coming days.

Kind regards,
Trixx

<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.www.avg.com
<http://www.avg.com/email-signature?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

Op zo 15 okt 2023 om 15:28 schreef Martin M. ***@***.***>:

**Comment by Trixx34** [comment]
@mcm1957, If you would be so kind to create a PR to fix issue #14, that would be amazing!
